### PR TITLE
New analytics models.

### PR DESCRIPTION
**Pull request by darrint**
## What's this PR do?

Adds some new analytics tables. Sort of. It won't actually work yet. I'm asking for some review.

## Where should the reviewer start?

The models.

## How should this be manually tested?

Nope.

## Any background context you want to provide?

na.

### CODE REVIEW

File: analytics/models.py [hunk 1]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    # jl=joblist
+    # js=jobsearch
+    # ja=jobapply
+    # jv=jobview
+    # ge=generic
+    # uk=unknown
+    page_type = models.CharField(max_length=255, null=True)
+
+    # `dn` - AKA "source"
+    campaign_name = models.CharField(max_length=255, null=True)
```
Comment (vigowei) [inline analytics/models.py]: remove this since we have `visit_campaign_name`

File: analytics/models.py [hunk 2]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    # generics is any other page, and unknown is anything that is not a page
+    # (redirects only, at this point)
+    # Choices:
+    # jl=joblist
+    # js=jobsearch
+    # ja=jobapply
+    # jv=jobview
+    # ge=generic
+    # uk=unknown
+    page_type = models.CharField(max_length=255, null=True)
```
Comment (darrint) [inline analytics/models.py]: Move to PageView

File: analytics/models.py [hunk 3]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    visit_campaign_term = models.CharField(max_length=255, null=True)
+    visit_campaign_content = models.CharField(max_length=255, null=True)
+
+    # Referrer when the visit began.
+    visit_referrer_hostname = models.CharField(max_length=255, null=True)
+
+    # Last goal event
+    goal_minus_one = models.CharField(max_length=255, null=True)
+    # Goal before last goal event
+    goal_minus_two = models.CharField(max_length=255, null=True)
```
Comment (vigowei) [inline analytics/models.py]: revisit, need to decide how many levels we need and who is responsible for maintaining the state

File: analytics/models.py [hunk 4]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    # uk=unknown
+    page_type = models.CharField(max_length=255, null=True)
+
+    # `dn` - AKA "source"
+    campaign_name = models.CharField(max_length=255, null=True)
+
+    # Referrer (or hostname) when this nv first started.
+    origin = models.CharField(max_length=255, null=True)
+    # Label based on consulting a table of hostnames/referrers.
+    channel = models.CharField(max_length=255, null=True)
```
Comment (vigowei) [inline analytics/models.py]: another stateful metric, it represents where the traffic is from. We are planning to put the hostname-referrer: channel pairs in Redis.

File: analytics/models.py [hunk 5]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    # hash_trunc(nv + job_guid)
+    job_view_id = models.BigIntegerField()
+
+    # Time this visit started
+    time_start = models.DateTimeField()
+    # sum of page view times so far.
+    time_on_site_msec = models.IntegerField(null=True)
+
+    # True when this visit is from a new myguid visitor (recent history)
+    flag_visitor = models.BooleanField()
```
Comment (vigowei) [inline analytics/models.py]: how long do we keep track of users to determine someone is new?

File: analytics/models.py [hunk 6]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    # True when this visit is from a new myguid visitor (all time)
+    flag_new_visitor = models.BooleanField()
```
Comment (vigowei) [inline analytics/models.py]: does a registered user have a fixed unique `myguid`?  i.e. it doesn't change from session to session

File: analytics/models.py [hunk 7]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+
+    page_views = models.IntegerField(null=True)
+
+    is_first_of_visit = models.BooleanField()
+    did_engage = models.BooleanField()
+    did_bounce = models.BooleanField()
```
Comment (vigowei) [inline analytics/models.py]: are those supposed to be the sum of pages engaged?  also, what signals a bounce?

File: analytics/models.py [hunk 8]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    time_on_site_msec = models.IntegerField(null=True)
+
+    # True when this visit is from a new myguid visitor (recent history)
+    flag_visitor = models.BooleanField()
+    # True when this visit is from a new myguid visitor (all time)
+    flag_new_visitor = models.BooleanField()
+
+    page_views = models.IntegerField(null=True)
+
+    is_first_of_visit = models.BooleanField()
```
Comment (vigowei) [inline analytics/models.py]: remove this. it belongs in PageView

File: analytics/models.py [hunk 9]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    time_start = models.DateTimeField()
+    # `et` from the last unload event, or null. Will stay null if we never
+    # get an unload event.
+    time_on_page_msec = models.IntegerField(null=True)
+
+    # Was this the first page view in a visit?
+    begins_visit = models.CharField(null=True)
+
+    # Is this the first time we saw this path in this visit?
+    is_new_within_visit = models.BooleanField()
```
Comment (vigowei) [inline analytics/models.py]: used to count unique page views.

File: analytics/models.py [hunk 10]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    job_view_id = models.BigIntegerField()
+
+    # `pn`
+    path = models.CharField(max_length=255, null=True)
+
+    # Time this page view started
+    time_start = models.DateTimeField()
+    # `et` from the last unload event, or null. Will stay null if we never
+    # get an unload event.
+    time_on_page_msec = models.IntegerField(null=True)
```
Comment (vigowei) [inline analytics/models.py]: we don't always get this info, it's None/null when not set, we can set it if and only if there is an unload event, by calculating elapsed time from the load event to the unload event.

File: analytics/models.py [hunk 11]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+
+    # Was this the first page view in a visit?
+    begins_visit = models.CharField(null=True)
+
+    # Is this the first time we saw this path in this visit?
+    is_new_within_visit = models.BooleanField()
+    # The user engaged during this page view.
+    did_engage = models.BooleanField()
+    # The user bounced.
+    did_bounce = models.BooleanField()
```
Comment (vigowei) [inline analytics/models.py]: how is this defined?? talk to Hal

File: analytics/models.py [hunk 12]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+
+class PageView(SiteTrafficBaseModel):
+    site_visit_id = models.BigIntegerField()
+    family_visit_id = models.BigIntegerField()
+    # hash_trunc(sv + aguid + page_guid)
+    page_view_id = models.BigIntegerField()
+    job_view_id = models.BigIntegerField()
+
+    # `pn`
+    path = models.CharField(max_length=255, null=True)
```
Comment (vigowei) [inline analytics/models.py]: should this field be included in the cluster index?

File: analytics/models.py (hunk 3 above)
Comment (darrint) [inline analytics/models.py]: Need Both per Hal.

File: analytics/models.py (hunk 6 above)
Comment (darrint) [inline analytics/models.py]: Yes. Per Hal.

File: analytics/models.py (hunk 5 above)
Comment (darrint) [inline analytics/models.py]: No. Count distinct aguids within hostname.

File: analytics/models.py (hunk 7 above)
Comment (darrint) [inline analytics/models.py]: Take this out, per Hal.

File: analytics/models.py [hunk 13]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+class Visit(SiteTrafficBaseModel):
+    # hash_trunc(sv + aguid)
+    site_visit_id = models.BigIntegerField()
+    # hash_trunc(fv + aguid)
+    family_visit_id = models.BigIntegerField()
+    # hash_trunc(nv + job_guid)
+    job_view_id = models.BigIntegerField()
+
+    # Time this visit started
+    time_start = models.DateTimeField()
```
Comment (darrint) [inline analytics/models.py]: REmove

File: analytics/models.py [hunk 14]
```diff
@@ -543,6 +544,146 @@ def save(self, *args, **kwargs):
...
+    # sum of page view times so far.
+    time_on_site_msec = models.IntegerField(null=True)
```
Comment (darrint) [inline analytics/models.py]: Time since first event of visit of the last event we received.

File: analytics/models.py (hunk 11 above)
Comment (darrint) [inline analytics/models.py]: Take out. Per Hal.

File: analytics/models.py (hunk 12 above)
Comment (darrint) [inline analytics/models.py]: No! We might group by page_type instead.

**Pull request closed without merging**